[NestedTensor] Support ragged_idx != 1 in pointwise ops

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118317
* __->__ #118157

This PR allows pointwise ops to operate on tensors with ragged_idx != 1. It does this by passing the ragged_idx metadata into the construction of the returned NestedTensor when computing pointwise ops. The assumption is that: pointwise ops can operate directly on the values tensors, and the resulting tensor should have all the same metadata properties as the input tensors. For binary ops, a test is added to verify that adding two tensors with different ragged_idx cannot be added.

Previously:
* unary pointwise ops would error out when performed on nested tensors with ragged_idx != 1
* binary pointwise ops would produce tensors with nonsense shapes

Differential Revision: [D53032641](https://our.internmc.facebook.com/intern/diff/D53032641)

davidberard98 committed:
[WIP][NestedTensor] Support ragged_idx != 1

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][NestedTensor] Support ragged_idx != 1"




[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D53032641).

davidberard98 committed:
Update on "[WIP][NestedTensor] Support ragged_idx != 1"


Differential Revision: [D53032641](https://our.internmc.facebook.com/intern/diff/D53032641)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D53032641).

jbschlosser commented on torch/nested/_internal/sdpa.py:
```diff
@@ -61,6 +61,11 @@ def _validate_sdpa_input(
             f"Expected query, key, and value to all be ragged on the same dimension, but got ragged "
             f"dims {query._ragged_idx}, {key._ragged_idx}, and {value._ragged_idx}, respectively."
         )
+    if query._ragged_idx != 1 and query._ragged_idx != 2:
```
What's the purpose of this check? I think it may be too restrictive and breaks if we have multiple batch dims:

e.g. input with shape `(B_0, B_1, S*, num_heads, head_dim)`
transposed to `(B_0, B_1, num_heads, S*, head_dim)` has `_ragged_idx = 3` and semantically we should support this.
```

davidberard98 replied:
Based on my reading of this function, the NT sdpa code seems unaware of the value of `_ragged_idx`? My intuition was that it would seem difficult to properly handle this correctly without looking at `_ragged_idx`, so I imagined that this must assume the ragged dimension is 2 (or 1 in the case of the special handling at the beginning of the fn).

Is this accurate?

I've removed the check for now though so we can discuss this separately :)

jbschlosser replied:
> I imagined that this must assume the ragged dimension is 2 (or 1 in the case of the special handling at the beginning of the fn)

The assumption SDPA makes is that the sequence length dim is at `dim=-2` (i.e. the second to last dim). This dim can either be constant or ragged, and it's handled differently between the two cases.

We don't support the embedding dim (`dim=-1`) being ragged, but any other dim can be ragged. If the sequence length dim is not the ragged one, we simply get a "ragged batch dim" or a ragged number of batch elements, which the underlying SDPA kernel doesn't care about.

davidberard98 replied:
Thanks for the explanation!

jbschlosser commented on torch/nested/_internal/ops.py:
```diff
@@ -204,6 +204,7 @@ def extract_kwargs(arg):
     kwargs = {
         "offsets": arg.offsets(),
         "_metadata_cache": arg._metadata_cache,
+        "_ragged_idx": arg._ragged_idx,
```
Note for posterity: this was left out here previously as `_ragged_idx != 1` was assumed to imply non-contiguity, and we generally don't preserve non-contiguity through out-of-place op invocations.

jbschlosser commented on test/test_nestedtensor.py:
```diff
@@ -3054,6 +3054,30 @@ def grad_test_func(a, b, c):
 
         gradcheck(grad_test_func, inputs=(a, b, c), check_batched_grad=False)
 
+    def test_unary_pointwise_transposed_inputs(self, device):
```
Thanks for the extensive testing! Do you mind also verifying that the contiguity is as expected in the new tests?

davidberard98 replied:
Added contiguity checks in the tests

jbschlosser commented on test/test_nestedtensor.py:
```diff
@@ -3555,9 +3614,11 @@ def test_noncontiguous_pointwise(self, device):
-        # pointwise ops are not supported on ragged dim transposed jagged layout NTs
-        with self.assertRaisesRegex(ValueError, "expected .* to be a contiguous jagged layout"):
-            clone = transposed.clone()
+        clone = transposed.clone()
```
please assert on contiguity here - we expect the clone to be contiguous but transposed not to be

davidberard98 replied:
I checked in eager non-NT; it seems like contiguity isn't actually preserved usually? I added tests for this, and also asserts that is_contiguity is true after `nt.transpose(1, 2).clone(memory_format=torch.contiguous_format)`

jbschlosser replied:
> I checked in eager non-NT; it seems like contiguity isn't actually preserved usually

that's right - my understanding is that regardless of input contiguity, `clone()` will spit out a contiguous tensor

davidberard98 replied:
> regardless of input contiguity, clone() will spit out a contiguous tensor

this was not what I observed (both with NT and eager) - usually it spits out a tensor with the same strides as the input, unless `contiguous_format` is specified

jbschlosser replied:
ah I see, thanks for clarifying!

jbschlosser reviewed (commented):
Thanks for the fix!

One thing I'm thinking through is ensuring the semantics are consistent with dense tensors in eager. In general, we have:
```
non-contiguous input -> out-of-place-op -> contiguous output
```
We should maintain this property for NT as well.

Previously, the only way to get `_ragged_idx != 1` was via `transpose()`, so such an NT would be necessarily non-contiguous. After this PR, we can get a NT with `_ragged_idx != 1` as the output of a unary op on such a non-contiguous NT input, and, by the property above, we expect it to be contiguous.

So we have to change the definition of `is_contiguous()` for jagged NT at least.

Another note: perf-wise, SDPA assumes that the inputs have been "pre-transposed" and that it can un-transpose them very cheaply. If we're sending the output of a unary op with `_ragged_idx != 1` to SDPA, this is no longer the case. No action needed for this PR per se; just pointing this out as something to consider.

YuqingJ commented:
> Thanks for the fix!
> 
> One thing I'm thinking through is ensuring the semantics are consistent with dense tensors in eager. In general, we have:
> 
> ```
> non-contiguous input -> out-of-place-op -> contiguous output
> ```
> 
> We should maintain this property for NT as well.
> 
> Previously, the only way to get `_ragged_idx != 1` was via `transpose()`, so such an NT would be necessarily non-contiguous. After this PR, we can get a NT with `_ragged_idx != 1` as the output of a unary op on such a non-contiguous NT input, and, by the property above, we expect it to be contiguous.
> 
> So we have to change the definition of `is_contiguous()` for jagged NT at least.
> 
> Another note: perf-wise, SDPA assumes that the inputs have been "pre-transposed" and that it can un-transpose them very cheaply. If we're sending the output of a unary op with `_ragged_idx != 1` to SDPA, this is no longer the case. No action needed for this PR per se; just pointing this out as something to consider.

Do you mean that the un-transpose in SDPA won't be cheap anymore after this PR?

jbschlosser commented:
> Do you mean that the un-transpose in SDPA won't be cheap anymore after this PR?

If there is a unary op invocation in between the transpose and the SDPA call, that's right.

davidberard98 committed:
Update on "[NestedTensor] Support ragged_idx != 1 in pointwise ops"


This PR allows pointwise ops to operate on tensors with ragged_idx != 1. It does this by passing the ragged_idx metadata into the construction of the returned NestedTensor when computing pointwise ops. The assumption is that: pointwise ops can operate directly on the values tensors, and the resulting tensor should have all the same metadata properties as the input tensors. For binary ops, a test is added to verify that adding two tensors with different ragged_idx cannot be added.

Previously:
* unary pointwise ops would error out when performed on nested tensors with ragged_idx != 1
* binary pointwise ops would produce tensors with nonsense shapes

Differential Revision: [D53032641](https://our.internmc.facebook.com/intern/diff/D53032641)

[ghstack-poisoned]

davidberard98 committed:
Update on "[NestedTensor] Support ragged_idx != 1 in pointwise ops"


This PR allows pointwise ops to operate on tensors with ragged_idx != 1. It does this by passing the ragged_idx metadata into the construction of the returned NestedTensor when computing pointwise ops. The assumption is that: pointwise ops can operate directly on the values tensors, and the resulting tensor should have all the same metadata properties as the input tensors. For binary ops, a test is added to verify that adding two tensors with different ragged_idx cannot be added.

Previously:
* unary pointwise ops would error out when performed on nested tensors with ragged_idx != 1
* binary pointwise ops would produce tensors with nonsense shapes

Differential Revision: [D53032641](https://our.internmc.facebook.com/intern/diff/D53032641)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D53032641).

jbschlosser reviewed (approved):
LGTM thanks!

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
@pytorchbot merge
